File: +ndi/+fun/name2variableName.m
function variableName = name2variableName(name)
%NAME2VARIABLENAME Converts a string into a camelCase variable name format.
%
%   variableName = name2variableName(name) takes a character array
%   (char), string array (string), or cell array of char vectors (cellstr)
%   as input and returns a new string suitable for use as a
%   variable name in MATLAB. This version primarily uses cellstr operations.
%
%   This involves:
%   1. Converting the input to a cell array of char vectors for consistent processing.
%   2. Replacing non-alphanumeric characters (except underscore) with spaces.
%   3. Splitting the cleaned string into individual words.
%   4. Capitalizing the first letter of each word and lowercasing the rest.
%   5. Joining all words together without spaces.
%   6. Converting the very first letter of the entire resulting string to lowercase.
%   7. Ensuring the string starts with a letter if it doesn't already
%      (by prepending 'x' if necessary).
%   8. Final cleanup to remove any remaining invalid characters.
%   9. Converting the output back to a char array if the original input was char.
%
%   Input:
%     name - The raw input string (char, string, or cellstr) to be converted.
%
%   Output:
%     variableName - The processed string formatted as a camelCase variable name.
%                   Returns a char array if input was char, otherwise a cellstr.
%

% Input argument validation
arguments
    name {mustBeText}
end

    % Store original input type to determine output type
    wasCharInput = ischar(name);
    wasStringInput = isstring(name); % To handle string array input type

    % Step 1: Ensure input is a cell array of char vectors for consistent processing.
    inputCell = cellstr(name);

    % Process each string in the input cell array (if it was a cellstr or string array)
    outputCell = cell(size(inputCell));

    for k = 1:numel(inputCell)
        currentInputStr = inputCell{k};

        % Handle empty string or string with only spaces for current element
        if isempty(currentInputStr) || all(isspace(currentInputStr))
            outputCell{k} = '';
            continue; % Skip to next element
        end

        % Step 2: Replace select characters with an underscore
        currentInputStr = replace(currentInputStr,':','_');
        currentInputStr = replace(currentInputStr,'-','_');

        % Step 3: Replace non-alphanumeric characters (except underscore) with spaces.
        cleanedChar = regexprep(currentInputStr, '[^a-zA-Z0-9_]', ' ');

        % Step 4: Split the cleaned string into words based on whitespace.
        % 'strsplit' returns a cell array of char vectors.
        words = strsplit(cleanedChar, ' ', 'CollapseDelimiters', true);

        % Initialize an empty cell array to store processed words
        capitalizedWords = cell(size(words));

        % Iterate through each word to capitalize its first letter
        for i = 1:numel(words)
            currentWord = words{i}; % Access as char vector

            % Check if the current word is not empty (e.g., from multiple spaces after cleaning)
            if ~isempty(currentWord)
                % Capitalize the first letter and convert the rest to lowercase
                firstLetter = upper(currentWord(1));
                if length(currentWord) > 1
                    restOfWord = lower(currentWord(2:end));
                else
                    restOfWord = ''; % Word has only one letter
                end
                capitalizedWords{i} = [firstLetter, restOfWord]; % Concatenate char arrays
            else
                capitalizedWords{i} = ''; % Keep empty char vector
            end
        end

        % Step 5: Join the capitalized words back together without spaces.
        tempVariableNameChar = strjoin(capitalizedWords, '');

        % Step 6: Ensure the variable name starts with a letter.
        if ~isempty(tempVariableNameChar) && ~isletter(tempVariableNameChar(1))
            finalVariableNameChar = ['var_', tempVariableNameChar];
        else
            finalVariableNameChar = tempVariableNameChar;
        end

        % Step 7: Final cleanup to remove any remaining non-alphanumeric characters.
        % This regex keeps only letters, numbers, and underscores.
        finalVariableNameChar = regexprep(finalVariableNameChar, '[^a-zA-Z0-9_]', '');

        outputCell{k} = finalVariableNameChar;
    end

    % Step 8: Convert the final output back to the original input type if applicable.
    if wasCharInput
        variableName = outputCell{1}; % If input was char, output is a single char array
    elseif wasStringInput
        variableName = string(outputCell); % If input was string array, convert back to string array
    else
        variableName = outputCell; % If input was cellstr, output is a cellstr
    end

end

File: ndi-python/ndi/fun/name2variablename.py
"""
name2variableName - Convert a string into a camelCase variable name format.

This module provides the name2variableName function which converts arbitrary
strings into valid Python variable names in camelCase format.
"""

import re
from typing import Union, List


def name2variableName(name: Union[str, List[str]]) -> Union[str, List[str]]:
    """
    Convert a string into a camelCase variable name format.

    This function takes a string or list of strings and converts them into
    valid variable names suitable for use in Python.

    Steps:
    1. Replace colons and hyphens with underscores
    2. Replace non-alphanumeric characters (except underscore) with spaces
    3. Split into words
    4. Capitalize the first letter of each word
    5. Join words together without spaces
    6. Ensure name starts with a letter (prepend 'var_' if needed)
    7. Final cleanup to remove invalid characters

    Args:
        name: The raw input string or list of strings to convert

    Returns:
        Processed string(s) formatted as camelCase variable name(s)

    Examples:
        >>> name2variableName('hello world')
        'HelloWorld'
        >>> name2variableName('CL:0000000')
        'CL_0000000'
        >>> name2variableName('test-name')
        'Test_Name'
        >>> name2variableName(['hello', 'world'])
        ['Hello', 'World']
    """
    # Handle list input
    if isinstance(name, list):
        return [name2variableName(n) for n in name]

    # Handle empty or whitespace-only strings
    if not name or not name.strip():
        return ''

    current_str = name

    # Step 1: Replace select characters with an underscore
    current_str = current_str.replace(':', '_')
    current_str = current_str.replace('-', '_')

    # Step 2: Replace non-alphanumeric characters (except underscore) with spaces
    cleaned = re.sub(r'[^a-zA-Z0-9_]', ' ', current_str)

    # Step 3: Split into words
    words = cleaned.split()

    # Step 4: Capitalize first letter of each word
    capitalized_words = []
    for word in words:
        if word:  # Non-empty word
            # Capitalize first letter, keep rest as-is
            capitalized = word[0].upper() + word[1:] if len(word) > 1 else word[0].upper()
            capitalized_words.append(capitalized)

    # Step 5: Join words together
    result = ''.join(capitalized_words)

    # Step 6: Ensure name starts with a letter
    if result and not result[0].isalpha():
        result = 'var_' + result

    # Step 7: Final cleanup - remove any remaining invalid characters
    result = re.sub(r'[^a-zA-Z0-9_]', '', result)

    return result
